packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -879,6 +904,16 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
         this.once("attaching", () => {
             assert(this.attachState === AttachState.Detached, 0x14d /* "Should move from detached to attaching" */);
             this._attachState = AttachState.Attaching;
+
+            assert(
+                this.visibilityState !== VisibilityState.GloballyVisible,
+                "Data store should not already be globally visible",
+            );
+            /**
+             * Once the data store moves to attaching state, its existence has been notified to the server which will
+             * notify all clients. The data store is now globally visible.
+             */
+            this.visibilityState = VisibilityState.GloballyVisible;
         });
         this.once("attached", () => {
             assert(this.attachState === AttachState.Attaching, 0x14e /* "Should move from attaching to attached" */);
```

Comment:
Maybe add assert here as well that this.visibilityState === VisibilityState.GloballyVisible ?

Reply:
I'd also consider adding similar invariant checks in public methods like generateAttachMessage()

Reply:
Added a similar assert to the `getAttachSummary` method in data store runtime.